File: 2020/day05/ex1.py
import sys

lines = [l.strip() for l in sys.stdin]
max_seat_id = 0
for line in lines:
    min_row = 0
    max_row = 127
    min_col = 0
    max_col = 7

    for c in line:
        if c == 'F':
            max_row = int(min_row + (max_row - min_row) / 2)
        elif c == 'B':
            min_row = int(min_row + (max_row - min_row + 1) / 2)
        elif c == 'L':
            max_col = int(min_col + (max_col - min_col) / 2)
        elif c == 'R':
            min_col = int(min_col + (max_col - min_col + 1) / 2)

    seat_id = min_row * 8 + min_col
    if seat_id > max_seat_id:
        max_seat_id = seat_id

print(max_seat_id)

File: 2020/day05/ex1.c
#include <stdio.h>

int main() {
  char buffer[64];
  int max_seat_id = 0;

  while (fgets(buffer, sizeof(buffer), stdin)) {
    char *c;
    int seat_id;
    int min_row = 0;
    int max_row = 127;
    int min_col = 0;
    int max_col = 7;

    for (c = buffer; *c && *c != '\n'; c++) {
      switch (*c) {
        case 'F':
            max_row = (int)(min_row + (max_row - min_row) / 2);
            break;
        case 'B':
            min_row = (int)(min_row + (max_row - min_row + 1) / 2);
            break;
        case 'L':
            max_col = (int)(min_col + (max_col - min_col) / 2);
            break;
        case 'R':
            min_col = (int)(min_col + (max_col - min_col + 1) / 2);
            break;
      }
    }

    seat_id = (min_row << 3) | min_col;
    if (seat_id > max_seat_id) {
      max_seat_id = seat_id;
    }
  }

  printf("%d\n", max_seat_id);

  return 0;
}
